cypher-shell/src/main/java/org/neo4j/shell/ShellRunner.java

```diff
@@ -40,13 +44,18 @@ static ShellRunner getShellRunner(@Nonnull CliArgs cliArgs,
                                       @Nonnull Logger logger) throws IOException {
         if (cliArgs.getCypher().isPresent()) {
             return new StringShellRunner(cliArgs, cypherShell, logger);
-        } else if (isInputInteractive()) {
-            return new InteractiveShellRunner(cypherShell, cypherShell, logger, new ShellStatementParser(),
-                    System.in, FileHistorian.getDefaultHistoryFile());
-        } else {
-            return new NonInteractiveShellRunner(cliArgs.getFailBehavior(), cypherShell, logger,
-                    new ShellStatementParser(), System.in);
         }
+
+        try {
+            if (cliArgs.getNonInteractive() || !isInputInteractive(System.getProperty("os.name"))) {
+                return new NonInteractiveShellRunner(cliArgs.getFailBehavior(), cypherShell, logger,
+                        new ShellStatementParser(), System.in);
+            }
+        } catch (UnsupportedOSException ignored) {
+        }
+        return new InteractiveShellRunner(cypherShell, cypherShell, logger, new ShellStatementParser(),
+                System.in, FileHistorian.getDefaultHistoryFile());
```

Comment:
I think I'd be happier if the logic here (checks and exception handling) was encapsulated in a single "should be interactive" method which just returned a boolean.